README
NAME
    Business::DK::CVR - Danish CVR (VAT Registration) code
    generator/validator

VERSION
    This documentation describes version 0.06 of Business::DK::CVR

SYNOPSIS
        use Business::DK::CVR qw(validate);

        my $rv;
        eval { $rv = validate(27355021); };

        if ($@) {
            die "Code is not of the expected format - $@";
        }

        if ($rv) {
            print "Code is valid";
        } else {
            print "Code is not valid";
        }

        #Using with Params::Validate
        #See also examples/
    
        use Params::Validate qw(:all);
        use Business::DK::CVR qw(validateCVR);
    
        eval {
            check_cvr(cvr => 27355021);
        };
    
        if ($@) {
            print "CVR is not valid - $@\n";
        }
    
        eval {
            check_cvr(cvr => 27355020);
        };
    
        if ($@) {
            print "CVR is not valid - $@\n";
        }
    
        sub check_cvr {
            validate( @_,
            { cvr =>
                { callbacks =>
                    { 'validate_cvr' => sub { validateCVR($_[0]); } } } } );
        
            print "$_[0] is a valid CVR\n";
    
        }

DESCRIPTION
    CVR is a company registration number used in conjuction with VAT
    handling in Denmark.

    If you want to use this module with Data::FormValidator please check:
    Data::FormValidator::Constraints::Business::DK::CVR

SUBROUTINES AND METHODS
  validate
    The function takes a single argument, a 10 digit CVR number.

    The function returns 1 (true) in case of a valid CVR number argument and
    0 (false) in case of an invalid CVR number argument.

    If the argument is a valid argument the sum is calculated by
    _calculate_sum based on the argument and the controlcifers array.

    The sum returned is checked using a modulus caluculation and based on
    its validity either 1 or 0 is returned.

  validateCVR
    Better name for export. This is just a wrapper for "validate"

  generate
    Generate is a function which generates valid CVR numbers, it is by no
    means an authority, since CVRs are generated and distributed by danish
    tax authorities, but it can be used to generate example CVRs for testing
    and so on.

PRIVATE FUNCTIONS
  _calculate_sum
    This function takes an integer and calculates the sum bases on the the
    controlcifer array.

EXPORTS
    Business::DK::CVR exports on request:

    "validate"
    "generate"
    "_calculate_sum"

DIAGNOSTICS
    *   The amount requested exceeds the maximum possible valid CVRs 9090908

        The number of valid CVRs are limited, so if the user requests a
        number of CVRs to be generated which exceeds the upper limit, this
        error is instantiated. See: "generate".

CONFIGURATION AND ENVIRONMENT
    The module requires no special configuration or environment to run.

DEPENDENCIES
    *   Params::Validate

    *   Exporter

    *   Carp

    *   Scalar::Util

    *   Class::InsideOut

    *   English

    *   Params::Validate

    *   Readonly

INCOMPATIBILITIES
    The module has no known incompatibilities.

BUGS AND LIMITATIONS
    The module has no known bugs or limitations.

TEST AND QUALITY
    Coverage of the test suite is at 100%

TODO
    *   Get the generate method thorougly tested

BUG REPORTING
    Please report issues via CPAN RT:

      http://rt.cpan.org/NoAuth/Bugs.html?Dist=Business-DK-CVR

    or by sending mail to

      bug-Business-DK-CVR@rt.cpan.org

SEE ALSO
    <http://www.cvr.dk/>
    Business::DK::PO
    Business::DK::CPR
    <http://search.cpan.org/dist/Algorithm-CheckDigits>
    <http://search.cpan.org/~mamawe/Algorithm-CheckDigits-0.38/CheckDigits/M
    11_008.pm>
    Data::FormValidator::Constraints::Business::DK::CVR

AUTHOR
    Jonas B. Nielsen, (jonasbn) - "<jonasbn@cpan.org>"

COPYRIGHT
    Business-DK-CVR is (C) by Jonas B. Nielsen, (jonasbn) 2006-2011

LICENSE
    Business-DK-CVR is released under the artistic license

    The distribution is licensed under the Artistic License, as specified by
    the Artistic file in the standard perl distribution
    (http://www.perl.com/language/misc/Artistic.html).



Changes
Change log for Business::DK::CVR

0.12 2020-06-02 Maintenance release, update not required

- Added missing specification of test requirement to cpanfile
  Thanks to Mohammad S Anwar (MANWAR) for the PR

0.11 2020-06-01 Maintenance release, update not required

- Improvements to Dist::Zilla config, only ExtUtils::MakeMaker supported via Dist::Zilla now. Module::Build support having been removed

- See the Article by Neil Bowers (NEILB) on the topic: https://neilb.org/2015/05/18/two-build-files-considered-harmful.html

- Thanks to Karen Etheridge (ETHER) for information and link to the mentioned article

0.10 2020-05-31 Maintenance release, update not required

- Migrating build system from Module::Build to Dist::Zilla

- Upgrading license from Artistic 1.0 to Artistic 2.0

- Thanks to Andreas König (ANDK) for the push to get this release uploaded

0.09 2014-08-23 Maintenance release, update not required

- Added specification of required Perl version, settling for
  5.8 for now as reported by Perl::Critic


0.08 2014-08-15 Maintenance release, update not required

- Addressing Kwalitee, aligning version for version consistency
  in modules in distribution

- Added changes.t automatic assertion of the integrity of the
  Changes file, see also BDKCVR-8 [BDKCVR-9]

  https://logiclab.jira.com/browse/BDKCVR-9


0.07 2013-08-01 Maintenance release, update not required

- Fixed up Changes file as part of my Questhub quest adhering to
  the standard described in: CPAN::Changes::Spec [BDKCVR-8]

  Quest: http://questhub.io/realm/perl/quest/51f0fd8e18ba7dcf4b000040
  Spec: https://metacpan.org/module/CPAN::Changes::Spec

  https://logiclab.jira.com/browse/BDKCVR-8

- Added contemporay boiler plate of Perl::Critic test [BDKCVR-5]

  https://logiclab.jira.com/browse/BDKCVR-5


0.06 2011-02 Maintenance release, update not required

- Added use of Params::Validate in key places and phased out use
  of Business::DK::PO

- Made control cifers Readonly

- The _length method has been removed, since it's use was
  obsoleted


0.05 2008-06-11 Feature release, update not required

- Add object oriented variant implementation [BDKCVR-1],
  introducing Class::Business::DK::CVR

- Added prerequisites test (Test::Prereq) [BDKCVR-2]

- Added requirement of Scalar::Util, we are stealing as much as
  we can from Michael Peters's distribution:

    Data::FormValidator::Constraints::DateTime

- Added a LICENSE file

- Added LICENSE file and split COPYRIGHT and LICENSE section in
  POD into two separate sections

- Restructured tests files to follow namespace in directory
  structure

- Added recursive_test_files to Build.PL

- Added create_readme to Build.PL

- Introduced generate method so valid CVRs now can be generated
  randomly

- Introduced
  Data::FormValidator::Constraints::Business::DK::CVR, this is
  sort of an example application, it is however just a simple
  wrapper fitting into the Data::FormValidator::Constraints
  framework - it is should be usable, not just a demo, but still
  experimental


0.04 2007-03-13 Maintenance release, update not required

- Fixed problem in t/critic.t, I had implemented a mix of my own
  and the recommended Perl::Critic test, so it gave unforseen
  test failures


0.03 2007-03-12 Maintenance release, update not required

- Added Perl::Critic test, t/critic.t and related t/perlcriticrc

- Updated to more contemporary versions of:
    t/pod-coverage.t
    t/pod.t

  These should be there for the quality, but will not be run as
  a part of the normal test run, unless the TEST_POD environment
  variable is set.

  This seem to be the defacto way of doing things.


0.02 2006-03-01 Maintenance release, update not required

- Added some more links to SEE ALSO section, apparently there is
  already a module on CPAN capable of validating CVR numbers
  this module does not use the name CVR so I did no locate it
  prior to releasing this module to CPAN


0.01 2006-02-21

- Initial version


lib/Business/DK/CVR.pm
package Business::DK::CVR;

use strict;
use warnings;
use vars qw($VERSION @EXPORT_OK);
use Carp qw(croak);
use Params::Validate qw(validate_pos SCALAR OBJECT ARRAYREF );
use Readonly;
use 5.008; #5.8.0

use base qw(Exporter);

$VERSION   = '0.12';
@EXPORT_OK = qw(validate validateCVR generate _calculate_sum);

use constant MODULUS_OPERAND => 11;
use constant MAX_CVRS        => 9090908;
use constant VALID           => 1;
use constant INVALID         => 0;

Readonly my @controlcifers => qw(2 7 6 5 4 3 2 1);

sub validateCVR {
    return validate(shift);
}

sub validate {
    my ($controlnumber) = @_;

    validate_pos( @_, { type => SCALAR, regex => qr/^\d{8}$/ } );

    my $sum = _calculate_sum( $controlnumber, \@controlcifers );

    if ( $sum % MODULUS_OPERAND ) {
        return INVALID;
    } else {
        return VALID;
    }
}

sub _calculate_sum {
    my ( $number, $controlcifers ) = @_;

    validate_pos( @_,
        { type => SCALAR, regex => qr/^\d+$/ },
        { type => ARRAYREF },
    );

    my $sum = 0;
    my @numbers = split //smx, $number;

    for ( my $i = 0; $i < scalar @numbers; $i++ ) {
        $sum += $numbers[$i] * $controlcifers->[$i];
    }
    return $sum;
}

sub generate {
    my @array = validate_pos( @_,
        { type => OBJECT | SCALAR, optional => 1 },
        { type => SCALAR, optional => 1, default => 1 },
        { type => SCALAR, optional => 1, default => 1 },
    );

    my ( $self, $amount, $seed ) = @array;

    if ( defined $self and $self =~ m/\d+/ ) {
        $seed   = $amount;
        $amount = $self;
    }

    my @cvrs;
    my $cvr;

    if ( $amount > MAX_CVRS ) {
        croak 'The amount requested exceeds the maximum possible valid CVRs ('
            . MAX_CVRS . ')';
    }

    my $count = $amount;
    while ($count) {
        $cvr = sprintf '%08d', $seed;
        if ( validate($cvr) ) {
            push @cvrs, $cvr;
            $count--;
        }
        $seed++;
    }

    if (wantarray) {
        return @cvrs;
    } else {
        if ( $amount == 1 ) {
            return $cvr;
        } else {
            return \@cvrs;
        }
    }
}

1;

__END__

=pod

=encoding utf8

=begin markdown

[![CPAN version](https://badge.fury.io/pl/Business-DK-CVR.svg)](http://badge.fury.io/pl/Business-DK-CVR)
[![Build Status](https://travis-ci.org/jonasbn/perl-business-dk-cvr.svg?branch=master)](https://travis-ci.org/jonasbn/perl-business-dk-cvr)
[![Coverage Status](https://coveralls.io/repos/jonasbn/bdkcvr/badge.png)](https://coveralls.io/r/jonasbn/bdkcvr)

=end markdown

=head1 NAME

Business::DK::CVR - Danish CVR (VAT Registration) number generator/validator

=head1 VERSION

This documentation describes version 0.12 of Business::DK::CVR

=head1 SYNOPSIS

    use Business::DK::CVR qw(validate);

    my $rv;
    eval { $rv = validate(27355021); };

    if ($@) {
        die "Code is not of the expected format - $@";
    }

    if ($rv) {
        print "Code is valid";
    } else {
        print "Code is not valid";
    }

    #Using with Params::Validate
    #See also examples/

    use Params::Validate qw(:all);
    use Business::DK::CVR qw(validateCVR);

    eval {
        check_cvr(cvr => 27355021);
    };

    if ($@) {
        print "CVR is not valid - $@\n";
    }

    eval {
        check_cvr(cvr => 27355020);
    };

    if ($@) {
        print "CVR is not valid - $@\n";
    }

    sub check_cvr {
        validate( @_,
        { cvr =>
            { callbacks =>
                { 'validate_cvr' => sub { validateCVR($_[0]); } } } } );

        print "$_[0] is a valid CVR\n";

    }

=head1 DESCRIPTION

CVR is a company registration number used in conjunction with VAT handling in
Denmark.

If you want to use this module with L<Data::FormValidator> please check:
L<Data::FormValidator::Constraints::Business::DK::CVR>

=head1 SUBROUTINES AND METHODS

=head2 validate

The function takes a single argument, a 10 digit CVR number.

The function returns 1 (true) in case of a valid CVR number argument and  0
(false) in case of an invalid CVR number argument.

If the argument is a valid argument the sum is calculated by B<_calculate_sum>
based on the argument and the control digits array.

The sum returned is checked using a modulus calculation and based on its
validity either 1 or 0 is returned.

=head2 validateCVR

Better name for export. This is just a wrapper for L</validate>

=head2 generate

Generate is a function which generates valid CVR numbers, it is by no means
an authority, since CVRs are generated and distributed by danish tax
authorities, but it can be used to generate example CVRs for testing and so on.

=head1 PRIVATE FUNCTIONS

=head2 _calculate_sum

This function takes an integer and calculates the sum bases on the the
control digits array.

=head1 EXPORTS

Business::DK::CVR exports on request:

=over

=item L</validate>

=item L</generate>

=item L</_calculate_sum>

=back

=head1 DIAGNOSTICS

=over

=item * The amount requested exceeds the maximum possible valid CVRs 9090908

The number of valid CVRs are limited, so if the user requests a number of CVRs
to be generated which exceeds the upper limit, this error is instantiated.
See: L</generate>.

=back

=head1 CONFIGURATION AND ENVIRONMENT

The module requires no special configuration or environment to run.

=head1 DEPENDENCIES

=over

=item * L<Params::Validate>

=item * L<Exporter>

=item * L<Carp>

=item * L<Scalar::Util>

=item * L<Class::InsideOut>

=item * L<English>

=item * L<Params::Validate>

=item * L<Readonly>

=back

=head1 INCOMPATIBILITIES

The module has no known incompatibilities.

=head1 BUGS AND LIMITATIONS

The module has no known bugs or limitations.

=head1 TEST AND QUALITY

Coverage of the test suite is at 100%

=head1 BUG REPORTING

Please report issue via GitHub

    https://github.com/jonasbn/perl-business-dk-cvr/issues

Alternatively report issues via CPAN RT:

  http://rt.cpan.org/NoAuth/Bugs.html?Dist=Business-DK-CVR

or by sending mail to

  bug-Business-DK-CVR@rt.cpan.org

=head1 SEE ALSO

=over

=item L<http://www.cvr.dk/>

=item L<Business::DK::CPR>, validation of Danish social security numbers

=item L<http://search.cpan.org/dist/Algorithm-CheckDigits>, an alternative implementation

=item L<http://search.cpan.org/~mamawe/Algorithm-CheckDigits-0.38/CheckDigits/M11_008.pm>, see above

=item L<Data::FormValidator::Constraints::Business::DK::CVR>, a validator for DFV, included in this distribution

=item L<Business::Tax::VAT>, for VAT rates and doing VAT calculations

=back

=head1 AUTHOR

Jonas B., (jonasbn) - C<< <jonasbn@cpan.org> >>

=head1 ACKNOWLEDGEMENTS

=over

=item * Karen Etheridge (ETHER)

=item * Andreas König (ANDK)

=item * Neil Bowers (NEILB)

=item * Mohammad S Anwar (MANWAR)

=back

=head1 COPYRIGHT

Business-DK-CVR and related is (C) by Jonas B., (jonasbn) 2006-2020

=head1 LICENSE

Business-DK-CVR is released under the Artistic License 2.0

=cut


Build.PL
# Build.PL for Business-DK-CVR

# $Id: Build.PL,v 1.3 2008-06-11 08:08:00 jonasbn Exp $

use Module::Build;
my $build = Module::Build->new(
    dist_author          => 'Jonas B. Nielsen, (jonasbn) <jonasbn@cpan.org>',
    module_name          => 'Business::DK::CVR',
    dist_version_from    => 'lib/Business/DK/CVR.pm',
    license              => 'artistic',
    create_makefile_pl   => 'traditional',
    create_readme        => 1,
    recursive_test_files => 1,
    requires             => {
        'Exporter'         => '0',       #core
        'Carp'             => '0',       #core
        'Scalar::Util'     => '0',
        'Class::InsideOut' => '0',
        'English'          => '0',
        'Params::Validate' => '0',
        'Readonly'         => '0',
        'perl'             => '5.8.0',
    },
    build_requires => {
        'Test::More'          => '0',    #core
        'Test::Exception'     => '0',
        'Test::Perl::Critic'  => '0',
        'Test::Prereq::Build' => '0',
        'Test::Taint'         => '0',
        'Test::Kwalitee'      => '0',
        'Taint::Runtime'      => '0',
        'Data::FormValidator' => '0',
    },
    meta_merge => {
        resources => {
            homepage => 'http://logicLAB.jira.com/browse/BDKCVR',
            bugtracker =>
                'http://rt.cpan.org/NoAuth/Bugs.html?Dist=Business-DK-CVR',
            repository => 'https://github.com/jonasbn/bdkcvr',
            ProjectChangelog =>
                'http://logiclab.jira.com/browse/BDKCVR#selectedTab=com.atlassian.jira.plugin.system.project%3Achangelog-panel',
            ProjectSite => 'http://logicLAB.jira.com/browse/BDKCVR',
            ProjectWiki =>
                'http://logiclab.jira.com/wiki/display/BDKCVR/Home',
        }
    },
);
$build->create_build_script();
